Source/Dafny/Util.cs

```diff
@@ -465,4 +467,666 @@ class IEnumerableComparer<T> : IEqualityComparer<IEnumerable<T>> {
       return hash.ToHashCode();
     }
   }
+
+  // Class to traverse a program top-down, especially aimed at identifying expressions and statements,
+  // and in their context
```

Comment:
The previous Top-down traversers were not fit for that purpose, because they would not either not traverse all expressions and statement, or they would traverse the same expressions that were in a ghost context and expressions that weren't in a ghost context. I needed a traverser that keeps context of where to traverse (e.g. for functions by method, etc.)